sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -0,0 +1,429 @@
+using System.Linq;
+using System.Text;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// An Azure Resource Manager resource identifier.
+    /// </summary>
+    public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, IComparable<ResourceIdentifier>
+    {
+        private const char Separator = '/';
+
+        private static ResourceType TenantResourceType = new ResourceType("Microsoft.Resources/tenants");
+        private static ResourceType SubscriptionResourceType = new ResourceType("Microsoft.Resources/subscriptions");
+        private static ResourceType ResourceGroupResourceType = new ResourceType("Microsoft.Resources/resourceGroups");
+        private static ResourceType ProviderResourceType = new ResourceType("Microsoft.Resources/providers");
+        private const string RootStringValue = "/";
+        private const string ProvidersKey = "providers";
+        private const string SubscriptionsKey = "subscriptions";
+        private const string LocationsKey = "locations";
+        private const string ResourceGroupsLowerKey = "resourcegroups";
+
+        private string? _stringValue;
+
+        /// <summary>
+        /// The root of the resource hierarchy.
+        /// </summary>
+        public static readonly ResourceIdentifier Root = new ResourceIdentifier(null, TenantResourceType, string.Empty, true);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ResourceIdentifier"/> class.
+        /// </summary>
+        /// <param name="resourceId"> The id string to create the ResourceIdentifier from. </param>
+        public ResourceIdentifier(string resourceId)
+        {
+            var id = Create(resourceId);
+            Init(id.Parent, id.ResourceType, id.Name);
+            ResourceType = id.ResourceType;
+            Name = id.Name;
+            Parent = id.Parent;
+            IsChild = id.IsChild;
+            _stringValue = resourceId;
+        }
+
+        private ResourceIdentifier(ResourceIdentifier? parent, ResourceType resourceType, string resourceName, bool isChild)
+        {
+            Init(parent, resourceType, resourceName);
+            ResourceType = resourceType;
+            Name = resourceName;
+            Parent = parent ?? Root;
+            IsChild = isChild;
+            if (parent is null)
+                _stringValue = RootStringValue;
+        }
+
+        private ResourceIdentifier(ResourceIdentifier parent, string resourceTypeName, string resourceName)
+            : this(parent, ChooseResourceType(resourceTypeName, parent), resourceName, true)
+        {
+        }
+
+        private ResourceIdentifier(ResourceIdentifier parent, string providerNamespace, string resourceTypeName, string resourceName)
+            : this(parent, new ResourceType(providerNamespace, resourceTypeName), resourceName, false)
+        {
+        }
+
+        private void Init(ResourceIdentifier? parent, ResourceType resourceType, string resourceName)
```

Comment:
If only some of the initialization is done here, would it make sense to scope the method name to what's happening?